opm/simulators/timestepping/AdaptiveTimeSteppingEbos.hpp

```diff
@@ -343,16 +350,50 @@ namespace Opm {
-                    if (solverVerbose_) {
-                        std::string msg;
-                        msg = causeOfFailure + "\nTimestep chopped to "
-                            + std::to_string(unit::convert::to(substepTimer.currentStepLength(), unit::day)) + " days\n";
-                        OpmLog::problem(msg);
-                    }
+                    const double minimumChoppedTimestep = 0.25 * unit::day;
+                    if (newTimeStep > minimumChoppedTimestep) {
+                        // Chop the timestep.
+                        substepTimer.provideTimeStepEstimate(newTimeStep);
+                        if (solverVerbose_) {
+                            std::string msg;
+                            msg = causeOfFailure + "\nTimestep chopped to "
+                                + std::to_string(unit::convert::to(substepTimer.currentStepLength(), unit::day)) + " days\n";
+                            OpmLog::problem(msg);
+                        }
+                        ++restarts;
+                    } else {
+                        // We are below the threshold, and will check if there are any
+                        // wells we should close rather than chopping again.
+                        std::set<std::string> failing_wells = consistentlyFailingWells(solver.model().stepReports());
+                        if (failing_wells.empty()) {
+                            // Found no wells to close, chop the timestep as above.
+                            substepTimer.provideTimeStepEstimate(newTimeStep);
+                            if (solverVerbose_) {
+                                std::string msg;
+                                msg = causeOfFailure + "\nTimestep chopped to "
+                                    + std::to_string(unit::convert::to(substepTimer.currentStepLength(), unit::day)) + " days\n";
+                                OpmLog::problem(msg);
+                            }
+                            ++restarts;
+                        } else {
+                            // Close all consistently failing wells.
+                            for (const auto& well : failing_wells) {
+                                solver.model().wellModel().forceShutWellByNameIfPredictionMode(well, substepTimer.simulationTimeElapsed());
+                            }
+                            substepTimer.provideTimeStepEstimate(dt);
+                            if (solverVerbose_) {
+                                std::string msg;
+                                msg = "\nProblematic well(s) were shut: ";
+                                for (const auto& well : failing_wells) {
+                                    msg += well;
+                                    msg += " ";
+                                }
+                                msg += "(retrying timestep)\n";
+                                OpmLog::problem(msg);
+                            }
+                        }
```

Comment:
why is not ++restart added here. This potentially leads to infinite simulations

Reply:
It should not lead to an infinite loop, because it will only enter the else branch if `failing_wells` is nonempty. In theory. Since there is something wrong here I'll look at it as soon as I can. My initial guess is that perhaps the `failing_wells` are nonempty, but all wells there are historical, and therefore none are actually shut, which would lead to an infinite loop. A simple workaround may be to add `if (solver.model().wellModel().hasTHPConstraints()) {...}` to the logic here as well.

Reply:
See discussion and proposed fix in #1671.